safehttp/plugins/fetch_metadata/fetch_metadata.go

```diff
@@ -0,0 +1,163 @@
+	for _, e := range endpoints {
+		m[e] = true
+	}
+	return &Plugin{corsProtected: m}
+}
+
+func (p *Plugin) resourceIsolationPolicy(r *safehttp.IncomingRequest) bool {
+	h := r.Header
+	if h.Get("Sec-Fetch-Site") != "cross-site" {
+		// The request is allowed to pass because one of the following applies:
+		// - Fetch Metadata is not supported by the browser
+		// - the request is same-origin, same-site or caused by the user
+		// explicitly interacting with the user-agent
+		return true
+	}
+
+	method := r.Method()
+	mode := h.Get("Sec-Fetch-Mode")
+	dest := h.Get("Sec-Fetch-Dest")
+	// Allow CORS options requests if neither Mode nor Dest is set.
+	// https://github.com/w3c/webappsec-fetch-metadata/issues/35
+	// https://bugs.chromium.org/p/chromium/issues/detail?id=979946
+	if mode == "" && dest == "" && method == safehttp.MethodOptions {
+		return true
+	}
+
+	if navigationalModes[mode] && navigationalDest[dest] && statePreservingMethods[method] {
+		// The request is cross-site, but a simple top-level navigation from a
+		// safe destination so we  allow it to pass.
+		return true
+	}
+	// The request is cross-site and not a simple navigation or from an unsafe
+	// destination so it is rejected.
+	return false
+}
+
+func (p *Plugin) navigationIsolationPolicy(r *safehttp.IncomingRequest) bool {
+	if h := r.Header; p.NavIsolation && h.Get("Sec-Fetch-Site") == "cross-site" && navigationalModes[h.Get("Sec-Fetch-Mode")] {
+		return false
+	}
+	return true
+}
+
+// SetReportOnly sets the Fetch Metadata policy mode to "report". This will
+// allow requests that violate the policy to pass, but will report the violation
+// using the RequestLogger. The method will panic if RequestLogger is nil.
+func (p *Plugin) SetReportOnly() {
+	if p.Logger == nil {
+		panic("logging service required for Fetch Metadata report mode")
+	}
+	p.reportOnly = true
+}
+
+// SetEnforce sets the Fetch Metadata policy mode to "enforce". This will reject
+// any requests that violates the policy provided by the plugin.
+func (p *Plugin) SetEnforce() {
+	p.reportOnly = false
+}
+
+// Before validates the safehttp.IncomingRequest using the Resource Isolation
+// Policy and, if enabled, the Navigation Isolation Policy. It only allows
+// requests to pass if they conform to the policy, if it's targeted to one of
+// the CORS-protected endpoints, specified when creating the plugin, or if the
+// mode is set to  "report", in which case the request is allowed to pass but
+// the  violation is reported. If a redirectURL was provided and the Navigation
+// Isolation Policy is enabled and fails, the IncomingRequest will be
+// redirected to redirectURL.
+func (p *Plugin) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	if p.corsProtected[r.URL.Path()] {
+		// The request is targeted to an endpoint on which Fetch Metadata
+		// policies are disabled because it is CORS-protected so we don't apply
+		// the policies.
+		return safehttp.Result{}
+	}
```

Comment:
Hm... Why is this information stored in a plugin at all?

Shouldn't this work like:

```
m := Mux()
m.Install(fetchmetadata.Plugin)
m.Handle(..., fetchmetadata.Disable())
```

?

Reply:
This will then be handled and tested after the per-mux configurations have been finalised and we have an orchestrator.

Reply:
Good call.